x-pack/auditbeat/processors/sessionmd/provider/kerneltracingprovider/kerneltracingprovider_linux.go

```diff
@@ -0,0 +1,524 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build linux && (amd64 || arm64) && cgo
+
+package kerneltracingprovider
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	quark "github.com/elastic/go-quark"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/x-pack/auditbeat/processors/sessionmd/provider"
+	"github.com/elastic/beats/v7/x-pack/auditbeat/processors/sessionmd/types"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type prvdr struct {
+	ctx            context.Context
+	logger         *logp.Logger
+	qq             *quark.Queue
+	qqMtx          *sync.Mutex
+	combinedWait   time.Duration
+	inBackoff      bool
+	backoffStart   time.Time
+	since          time.Time
+	backoffSkipped int
+	mtx            *sync.Mutex
+}
+
+type TTYType int
+
+const (
+	TTYUnknown TTYType = iota
+	Pts
+	TTY
+	TTYConsole
+)
+
+const (
+	Init         = "init"
+	Sshd         = "sshd"
+	Ssm          = "ssm"
+	Container    = "container"
+	Terminal     = "terminal"
+	Kthread      = "kthread"
+	EntryConsole = "console"
+	EntryUnknown = "unknown"
+)
+
+const (
+	ptsMinMajor     = 136
+	ptsMaxMajor     = 143
+	ttyMajor        = 4
+	consoleMaxMinor = 63
+	ttyMaxMinor     = 255
+)
+
+var (
+	bootID     string
+	pidNsInode uint64
+)
+
+func readBootID() (string, error) {
+	bootID, err := os.ReadFile("/proc/sys/kernel/random/boot_id")
+	if err != nil {
+		return "", fmt.Errorf("could not read /proc/sys/kernel/random/boot_id, process entity IDs will not be correct: %w", err)
+	}
+
+	return strings.TrimRight(string(bootID), "\n"), nil
+}
+
+func readPIDNsInode() (uint64, error) {
+	var ret uint64
+
+	pidNsInodeRaw, err := os.Readlink("/proc/self/ns/pid")
+	if err != nil {
+		return 0, fmt.Errorf("could not read /proc/self/ns/pid, process entity IDs will not be correct: %w", err)
+	}
+
+	if _, err = fmt.Sscanf(pidNsInodeRaw, "pid:[%d]", &ret); err != nil {
+		return 0, fmt.Errorf("could not parse contents of /proc/self/ns/pid (%s)process entity IDs will not be correct: %w", pidNsInodeRaw, err)
+	}
+
+	return ret, nil
+}
+
+func NewProvider(ctx context.Context, logger *logp.Logger) (provider.Provider, error) {
+	attr := quark.DefaultQueueAttr()
+	attr.Flags = quark.QQ_ALL_BACKENDS | quark.QQ_ENTRY_LEADER | quark.QQ_NO_SNAPSHOT
+	qq, err := quark.OpenQueue(attr, 64)
+	if err != nil {
+		return nil, fmt.Errorf("open queue: %w", err)
+	}
+
+	var qqMtx sync.Mutex
+	var mtx sync.Mutex
+	p := prvdr{
```

Comment:
We know this is going to escape to the heap so might as well be direct about it.


`p := &prvdr{`